Pass stream ID arrays when calling peerConnection.addTrack and addTransceiver

csheely opened the pull request:
closes #546 

This pull request contains our proposal to supply media stream IDs from the javascript peer connection addTrack() and addTransceiver() methods.  IDs are passed in as an array, which is similar to the underlying WebRTC API.

sgodin committed:
-some proposed changes for modifying 2 node-webrtc to more closely match C++ versions
-addTrack modified from taking a MediaStream pointer to a taking in a vector of strings / media stream ids
-addTransceiver modified from taking a RtpTransceiverInit property that contains a vector of media streams,
 to one that takes a vector of strings / media stream ids
-attempted to modify the required test and javascript files
  -including the marking some potential additional work with TODOs in rtcrtpsender.js

csheely committed:
Fix typo and convert from Maybe<> obj to real obj to resolve compiler error

csheely committed:
Modify addTrack tests to pass ID array instead of stream object

csheely committed:
Reimplement addTransceiver test, but test may be unnecessary now

csheely committed:
Remove unused variable

sgodin committed:
-added a new MediaStream constructor to allow passing of a custom media stream id via new RTCMediaStreamInit struct
  -added a unit test for this new constructor to test/mediastream.js
-modified PeerConnection AddTrack to accept a vector of MediaStream's
-reverted previous changes to RtpTransceiverInit so we go back to accepting a vector of MediaStreams
-enabled rtcrtpsender.js test in all.js test suite
-adjusted test/rtcrtpsender.js for new addTrack signature that now accepts a vector of MediaStreams
-corrected test/rtcrtpsender.js expected test results for removeTrack

sgodin committed:
-update with changes from node-webrtc:develop master

sgodin committed:
Merge branch 'develop' into develop

sgodin committed:
-added missing semi-colon

sgodin committed:
-fix up destructor/index.js tests - addTrack takes an array now

markandrus commented on lib/peerconnection.js:
```diff
@@ -204,8 +204,8 @@ RTCPeerConnection.prototype.addTransceiver = function addTransceiver() {
-RTCPeerConnection.prototype.addTrack = function addTrack(track, stream) {
-  return this._pc.addTrack(track, stream);
+RTCPeerConnection.prototype.addTrack = function addTrack(track, streams) {
+  return this._pc.addTrack(track, streams);
```
Please check out the Web IDL for `addTrack` [here](https://w3c.github.io/webrtc-pc/#rtcpeerconnection-interface-extensions) — `streams` should use the [rest parameter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters) syntax. So, at least in the JavaScript layer here, we should define this method as

```js
RTCPeerConnection.prototype.addTrack = function addTrack(track, ...streams) {
  return this._pc.addTrack(track, streams);
};
```

node-webrtc supports Node 8 onwards, so I believe we ca safely use rest parameter syntax here.

sgodin replied:
Will do.

markandrus commented on src/dictionaries/node_webrtc/rtc_media_stream_init.cc:
```diff
@@ -0,0 +1,18 @@
+
+#define DICT(X) RTC_MEDIA_STREAM_INIT ## X
+#include "src/dictionaries/macros/impls.h"
+#undef DICT
```
👍

markandrus commented on src/dictionaries/node_webrtc/rtc_media_stream_init.h:
```diff
@@ -0,0 +1,16 @@
+#define DICT(X) RTC_MEDIA_STREAM_INIT ## X
+#include "src/dictionaries/macros/def.h"
+#include "src/dictionaries/macros/decls.h"
+#undef DICT
```
👍

markandrus commented on test/destructor/index.js:
```diff
@@ -109,9 +109,9 @@ test('Destructors fire in MediaStreamTrack use-case', async t => {
     const [pc1, pc2] = await negotiateRTCPeerConnections({
       withPc1(pc1) {
-        stream1.getTracks().forEach(track => pc1.addTrack(track, stream1));
+        stream1.getTracks().forEach(track => pc1.addTrack(track, [stream1]));
```
Per comment above, using rest parameter syntax, there should be no change necessary here.

markandrus commented on test/mediastream.js:
```diff
@@ -75,6 +75,13 @@ tape('new MediaStream(tracks)', function(t) {
 });
 
+tape('new MediaStream(mediaStreamInit)', function(t) {
+  var stream = new MediaStream({ id: 'testStreamId' });
```
Ca you please include a test to see what happens when two MediaStreams with the same ID are created? Also, I'm curious what happens if two MediaStreams with the same ID are used on an RTCPeerConnection.

sgodin replied:
Sorry I should have included some details on this.  I did set out to do this, however I backed it out.  Here's what I did:
-I analyzed the webrtc.org code base for what would happen if 2 MediaStreams had the same id.  I discovered that there was no major concern here.  It appears that worse case the 2nd stream would fail to get added to a peer connection if one with the same name already existed.
-I thought about adding a test case to test/mediastream.js - however this file really just deals with MediaStream objects in isolation.  So even though I could create test case where we create 2 media streams with the same id, we wouldn't really be testing anything useful, until they are added to a track or transceiver.
-I then set out to add such test to test/rtcrtpsender.js (since it has addTrack tests in it), however I was unable to come up with any meaningful way to verify anything (ie: assert conditions), since I can't actually retrieve the stream id, from an interface outside the MediaStream object itself, once it is set.  Also - we know that addTrack and addTransceiver in the C++ side don't really do anything with the media stream object, other than extracting the id from it.

If you have some other idea's or approaches I could use, I would be happy to try to implement them.  :)

Scott

sgodin replied:
OK, I found a way to verify things by looking at the generated SDP.  I added a new test case to rtcrtpsender.js - let me know what you think.  :)

markandrus commented on test/rtcrtpsender.js:
```diff
@@ -43,45 +42,45 @@ var sdp = [
 ].join('\r\n') + '\r\n';
 
-tape('.addTrack(track, stream)', function(t) {
+tape('.addTrack(track, [stream])', function(t) {
```
Per comment above, using rest parameter syntax, there should be no change necessary here. Can you please revert all such changes below?

markandrus reviewed (commented):
Overall looks good. Thanks for y'all's help here. There is one set of changes we still need related to rest parameter syntax in the JavaScript layer. Once that is completed, I believe we can merge.

sgodin committed:
-use the rest parameter syntax for addTrack
-reverted some changes to test/destructor/index.js and rtcrtpsender.js since array is no longer needed with rest syntax
-added test case to rtcrtpsender.js for multiple streams in addTrack and duplicate stream id's

sgodin committed:
-fix line endings

markandrus reviewed (approved):
Thanks so much!!!
